`DocumentArray` constructor accepts array of any type what can cause runtime error

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

6.10.0

### Node.js version

18.12.0

### MongoDB server version

N/A

### Typescript version (if applicable)

4.9.5

### Description

`DocumentArray` constructor [accepts](https://github.com/Automattic/mongoose/blob/6.10.0/types/types.d.ts#L64) `values: any[]`:
```ts
class DocumentArray<T> extends Types.Array<T extends Types.Subdocument ? T : Types.Subdocument<InferId<T>> & T> {
    /** DocumentArray constructor */
    constructor(values: any[]);
```
This can result in runtime error when passed to constructor value won't have required field.

### Steps to Reproduce

This code compiles and results in runtime error:
```ts
import { Types } from 'mongoose';

interface Book {
    author: {
        name: string;
    };
}

const books: Types.DocumentArray<Book> = new Types.DocumentArray<Book>([1, 2, 3]);
console.log(books[0].author.name);
```
Error:
```
console.log(books[0].author.name);
                            ^
TypeError: Cannot read properties of undefined (reading 'name')
```

### Expected Behavior

Compile time error:
```
Type 'number' is not assignable to type 'Book'
```

We'll work on a better workaround for this, but for now you can add `default: () => ({})` to the `location` property as follows.

```ts
const pointSchema = new Schema({
    name: { required: true, type: String },
    location: { required: true, type: locationSchema, default: () => ({}) }, // <-- `default` here
});
```

That gives you the TypeError that you expect:

```
$ ./node_modules/.bin/tsc gh-13087.ts 
gh-13087.ts:31:9 - error TS2322: Type 'DocumentArray<{ name: string; }>' is not assignable to type 'DocumentArray<{ name: string; location: { type: "Point"; coordinates: number[]; }; }>'.
  Property 'location' is missing in type '{ name: string; }' but required in type '{ name: string; location: { type: "Point"; coordinates: number[]; }; }'.

31         points: new Types.DocumentArray([

```

--- a/test/types/docArray.test.ts
+++ b/test/types/docArray.test.ts
@@ -1,4 +1,4 @@
-import { Schema, model, Document, Types } from 'mongoose';
+import { Schema, model, Types, InferSchemaType } from 'mongoose';
 import { expectError, expectType } from 'tsd';
 
 async function gh10293() {
@@ -23,4 +23,48 @@ async function gh10293() {
     expectType<string[][]>(test.arrayOfArray);
     return test.arrayOfArray; // <-- error here if the issue persisted
   };
+}
+
+function gh13087() {
+  interface Book {
+    author: {
+      name: string;
+    };
+  }
+
+  expectError(new Types.DocumentArray<Book>([1, 2, 3]));
+
+  const locationSchema = new Schema(
+    {
+      type: {
+        required: true,
+        type: String,
+        enum: ['Point']
+      },
+      coordinates: {
+        required: true,
+        type: [Number]
+      }
+    },
+    { _id: false }
+  );
+
+  const pointSchema = new Schema({
+    name: { required: true, type: String },
+    location: { required: true, type: locationSchema }
+  });
+
+  const routeSchema = new Schema({
+    points: { type: [pointSchema] }
+  });
+
+  type Route = InferSchemaType<typeof routeSchema>;
+
+  expectError(function getTestRouteData(): Route {
+    return {
+      points: new Types.DocumentArray([
+        { name: 'Test' } // "location" is missing
+      ])
+    };
+  });
 }

--- a/test/types/docArray.test.ts
+++ b/test/types/docArray.test.ts
@@ -51,7 +51,7 @@ function gh13087() {
 
   const pointSchema = new Schema({
     name: { required: true, type: String },
-    location: { required: true, type: locationSchema }
+    location: { required: true, type: String }
   });
 
   const routeSchema = new Schema({

--- a/test/types/docArray.test.ts
+++ b/test/types/docArray.test.ts
@@ -33,38 +33,4 @@ function gh13087() {
   }
 
   expectError(new Types.DocumentArray<Book>([1, 2, 3]));
-
-  const locationSchema = new Schema(
-    {
-      type: {
-        required: true,
-        type: String,
-        enum: ['Point']
-      },
-      coordinates: {
-        required: true,
-        type: [Number]
-      }
-    },
-    { _id: false }
-  );
-
-  const pointSchema = new Schema({
-    name: { required: true, type: String },
-    location: { required: true, type: String }
-  });
-
-  const routeSchema = new Schema({
-    points: { type: [pointSchema] }
-  });
-
-  type Route = InferSchemaType<typeof routeSchema>;
-
-  expectError(function getTestRouteData(): Route {
-    return {
-      points: new Types.DocumentArray([
-        { name: 'Test' } // "location" is missing
-      ])
-    };
-  });
 }

--- a/types/types.d.ts
+++ b/types/types.d.ts
@@ -61,7 +61,7 @@ declare module 'mongoose' {
 
     class DocumentArray<T> extends Types.Array<T extends Types.Subdocument ? T : Types.Subdocument<InferId<T>> & T> {
       /** DocumentArray constructor */
-      constructor(values: any[]);
+      constructor(values: (AnyKeys<T> & AnyObject)[]);
 
       isMongooseDocumentArray: true;
 

--- a/types/types.d.ts
+++ b/types/types.d.ts
@@ -61,7 +61,7 @@ declare module 'mongoose' {
 
     class DocumentArray<T> extends Types.Array<T extends Types.Subdocument ? T : Types.Subdocument<InferId<T>> & T> {
       /** DocumentArray constructor */
-      constructor(values: (AnyKeys<T> & AnyObject)[]);
+      constructor(values: (T & AnyObject)[]);
 
       isMongooseDocumentArray: true;
 

--- a/types/types.d.ts
+++ b/types/types.d.ts
@@ -61,7 +61,7 @@ declare module 'mongoose' {
 
     class DocumentArray<T> extends Types.Array<T extends Types.Subdocument ? T : Types.Subdocument<InferId<T>> & T> {
       /** DocumentArray constructor */
-      constructor(values: (T & AnyObject)[]);
+      constructor(values: (AnyKeys<T> & AnyObject)[]);
 
       isMongooseDocumentArray: true;
 

